ImportSource changes

kerimoyle opened the pull request:
Demonstrates #652

kerimoyle committed:
Merge pull request #41 from cellml/develop

Update

kerimoyle committed:
Merge pull request #42 from cellml/develop

Update from upstream

kerimoyle committed:
Merge pull request #43 from cellml/develop

Update from upstream

kerimoyle committed:
test to show bug

kerimoyle committed:
import sources tracked in model, parse bug not fixed yet

kerimoyle committed:
tests for import units

kerimoyle committed:
added storage to the importsource for dependent entities

kerimoyle committed:
printer fixed, other stuff broken ...

kerimoyle committed:
WIP unit tests passing (needs docstrings)

kerimoyle commented:
This is a work in progress.  I've changed:
- the Model keeps track of all ImportSources that it needs,
- the ImportSources keep track of the things which depend on them,
- the Printer bug has been fixed so that all imports which share a source are grouped together.  

I need to know whether there's a need to separate imports for any reason, and what a mechanism for that could be.  At present, if they have the same url but different ids then they're not considered the same (so that annotation is preserved).  Is there more that should be done?

All the unit tests pass, but I've changed two that required printing imports separately.  They'll fail here as I haven't done the bindings/coverage yet.  Thoughts/advice welcome ...

nickerso commented on tests/model/component_import.cpp:
```diff
@@ -201,14 +188,14 @@ TEST(ComponentImport, multipleImportAndParse)
     libcellml::PrinterPtr printer = libcellml::Printer::create();
     std::string a = printer->printModel(m);
-    EXPECT_TRUE((e1 == a) || (e2 == a));
+    EXPECT_EQ(e, a);
```
This is wrong, `c3` should be a in a separate import to `c1` and `c2`.

kerimoyle replied:
Because it was created with separate ImportSource pointers? Or is there another way to tell?

nickerso replied:
yep, that is the only way to tell I think

agarny replied:
Agreed and that means that we will somehow need to keep track of that fact when parsing a CellML file.

kerimoyle replied:
OK, I'll take out the check-for-duplicate-url bit, easy enough.

kerimoyle replied:
Done in fca5344

nickerso commented on tests/model/component_import.cpp:
```diff
@@ -201,14 +188,14 @@ TEST(ComponentImport, multipleImportAndParse)
     EXPECT_EQ(size_t(3), model->componentCount());
     a = printer->printModel(model);
-    EXPECT_TRUE((e1 == a) || (e2 == a));
+    EXPECT_EQ(e, a);
```
This is correct, as you get back the same model as parsed, but this is a different model to the one programmatically constructed above...

kerimoyle replied:
Done in fca5344

kerimoyle committed:
Changed functionality as requested: duplicated import sources possible

kerimoyle committed:
Added test for parser too

kerimoyle committed:
Update importsource.cpp

kerimoyle committed:
Typo in docstrings def

kerimoyle committed:
docstrings and convenience methods; coverage to follow.

kerimoyle commented:
I've reinstated the storage of duplicate import sources and changed the tests as appropriate.  Note that this means we can no longer retrieve an import by its `url` as it's not unique any more.  My thoughts at the moment are:
- In this PR we're keeping imported items separate, which seems to be is in opposition to what the Importer is doing in https://github.com/cellml/libcellml/pull/608? Should the importer also be importing things multiple times too? 
- I'd like to provide a convenience method so that if people have created multiple duplicates of imports, but don't actually need them, that the imports can be collapsed.  Since we've previously (erroneously) been printing individual units imports, and making them separate from all component imports, I'm guessing that the PMR is full of unintended duplicates ... Any thoughts?

(ps - I've left in lots of commented sections, will remove when I'm sure they're for the bin)

Edit: Other thoughts:
- When adding/removing a component/units item from an import source, should the import source be also added/removed from that component/units item in return? ie: a two-way relationship should be maintained?  The downstream effects are that removing a component from an import source will set it to return `isImport()` false until a replacement import source is added.

kerimoyle committed:
Some tests for coverage

kerimoyle committed:
more coverage tests

kerimoyle committed:
Added two way relationship between import and source

kerimoyle committed:
Add weak pointer for import source

kerimoyle committed:
Merge branch 'develop' into i652_parse_print_duplicate_ids

kerimoyle commented:
Ignore this for now ... I'm rethinking how it should be done.  Thoughts on users' required/expected functionality welcome though :)

kerimoyle committed:
Reworked storage of imported entities, tests and coverage pass

kerimoyle committed:
Merge branch 'i652_parse_print_duplicate_ids' of https://github.com/kerimoyle/libcellml into i652_parse_print_duplicate_ids

kerimoyle committed:
parser to add import source to model too

kerimoyle committed:
adding separate tests for import source to find memory leaks

kerimoyle committed:
more memory tests

kerimoyle committed:
Test to show bug when resetting import source

kerimoyle committed:
fix for bug in previous commit

kerimoyle committed:
Removed unreachable code, tweak tests

kerimoyle commented:
Hi @agarny @hsorby @nickerso As far as I can tell (without the CI build working properly) this is ready for review please.  It builds, passes tests and coverage on my machine.
Things to note:
- Where separate `<import>` blocks are parsed or `ImportSource` objects added using the API which use the same source URL, these are kept separate.  I will add a `consolidate()` or similar function to the `Importer` class as there will be loads of models out there which, because of the printer bug, have unintended separate import sources.
- Where different import children (components/units) exist in a single `ImportSource`, these are kept together (fixing the bug that this PR started from).
- When a mixture of components and units are in the same import source, they are printed in order of components first and then units.  
- An empty import source is printed (because it's valid according to https://cellml-specification.readthedocs.io/en/latest/reference/formal_and_informative/specB02.html )

kerimoyle committed:
test to show printer ordering

kerimoyle committed:
consolidate imports function (in model class)

kerimoyle committed:
Merge branch 'develop' into i652_parse_print_duplicate_ids

kerimoyle committed:
Added removeAll functions for Component, Units, ImportSource

kerimoyle committed:
Merge branch 'i652_parse_print_duplicate_ids' into consolidate_imports

kerimoyle committed:
WIP missing docstrings

kerimoyle committed:
Merge branch 'develop' into i652_parse_print_duplicate_ids

kerimoyle committed:
Merge branch 'i652_parse_print_duplicate_ids' of https://github.com/kerimoyle/libcellml into i652_parse_print_duplicate_ids

kerimoyle committed:
coverage and docstrings

kerimoyle committed:
formatting

kerimoyle commented:
Booo ... the CI is building based on an outdated commit again :( @hsorby

kerimoyle commented:
I've now included a `model->consolidateImports()` function to this PR.  The motivation for that is the bug in the `Printer` will have created a multitude of separate import sources throughout the PMR, and I wanted to give users an easy way to rationalise them.  I've played with the idea of moving this into the `Importer` class, but since it doesn't actually need anything to _be_ imported, I decided to leave it in the `Model` class.  Happy to move it if there's objection, but that would require this PR and #608 to be merged first.

nickerso commented:
> I've now included a `model->consolidateImports()` function to this PR. 

This doesn't seem like useful API and I'd suggest not including it.

kerimoyle commented:
> This doesn't seem like useful API and I'd suggest not including it.

Over time the number of import groups will have multiplied because of the printer bug ... this is just a way to let people fix that (easily) (if they want to).  I don't really understand the objection @nickerso? Too many bells and whistles or something else?

kerimoyle committed:
Merge pull request #44 from cellml/develop

Update from upstream

nickerso commented:
The only possible use case for this API is to "fix" a bug in code that realistically no one is using to produce CellML models. There is no way to implement a "consolidate imports" method that will avoid future confusion and introduction of buggy models as it seems like a perfectly reasonable method to call that will have loads of unintended consequences.

kerimoyle commented:
> The only possible use case for this API is to "fix" a bug in code that realistically no one is using to produce CellML models. 

Fair enough, I can change it back.

> There is no way to implement a "consolidate imports" method that will avoid future confusion and introduction of buggy models as it seems like a perfectly reasonable method to call that will have loads of unintended consequences.

... I don't follow this bit though? They only get merged when the URL is identical, so there should be no effect on the meaning of the instantiated model.  I guess it doesn't matter since I'll remove it anyway, but I like to understand what's going on (where possible ...) :)

kerimoyle committed:
removed consolidateImports functionality, added docstrings

nickerso commented:
> I don't follow this bit though? They only get merged when the URL is identical, so there should be no effect on the meaning of the instantiated model. I guess it doesn't matter since I'll remove it anyway, but I like to understand what's going on (where possible ...) :)

Two components in separate imports with the same URL is different to one import with two components in it. These are the types of differences that I don't think we are really testing at present.

kerimoyle commented:
> Two components in separate imports with the same URL is different to one import with two components in it. 

Wow, ok, I didn't know that!  How is it different?  Because we treat them exactly the same?  And we print them out together? And ... etc??  The only difference I can imagine is having different id attributes, but that doesn't fundamentally affect the semantics of the model itself?

nickerso commented on src/importedentity.cpp:
```diff
@@ -18,6 +18,8 @@ limitations under the License.
 
 namespace libcellml {
 
+// using ImportSourceWeakPtr = std::weak_ptr<ImportSource>;
```
this can be removed?

kerimoyle replied:
Probably :)

kerimoyle replied:
6e455f8c

nickerso commented on src/importedentity.cpp:
```diff
@@ -43,11 +45,13 @@ ImportedEntity::~ImportedEntity()
 
 bool ImportedEntity::isImport() const
 {
+    // return !mPimpl->mImportSource.expired();
```
random question, but these changes are due to the new use of `shared_from_this` for the import sources...or something, right?

kerimoyle replied:
Yes.  We now have a strong connection between `ImportedEntity` items and their `ImportSource`, and a weak connection between the `Model` and the `ImportSource`s it contains.

kerimoyle replied:
I've removed the comment in 6e455f8c

nickerso commented on src/parser.cpp:
```diff
@@ -126,13 +126,13 @@ struct Parser::ParserImpl
      * with names matching those in @p node, they will be overwritten. Likewise,
      * any imported components or units of the same name already in @p model will
-     * be overwritten by those parsed from @p node.
+     * be overwritten by those parsed from @p node.cmake
```
`cmake` ???

kerimoyle replied:
Ooops! Typed in the wrong window I guess ... 6e455f8c

nickerso commented on tests/import_source/import_source.cpp:
```diff
@@ -0,0 +1,400 @@
+
+TEST(ImportSource, createImportSource)
+{
+    auto imp1 = libcellml::ImportSource::create();
```
shouldn't there be an `EXPECT...` something here?

kerimoyle replied:
Yes 6e455f8c

nickerso commented on tests/model/component_import.cpp:
```diff
@@ -201,14 +191,14 @@ TEST(ComponentImport, multipleImportAndParse)
     libcellml::PrinterPtr printer = libcellml::Printer::create();
     std::string a = printer->printModel(m);
-    EXPECT_TRUE((e1 == a) || (e2 == a));
+    EXPECT_EQ(e, a);
```
just to check my understanding, because we now track imports in the models there is no longer any ambiguity in the order in which these get printed and hence no need to check both ordering options?

kerimoyle replied:
Yep, that's it.

nickerso commented on tests/printer/printer.cpp:
```diff
@@ -295,3 +295,52 @@ TEST(Printer, printModelWithStandardUnitsAdded)
+}
+
+TEST(Printer, printMultipleChildrenOfImportsMixedOrder)
+{
```
interesting test :)

kerimoyle replied:
I ran out of interestingness ... :)

nickerso reviewed (dismissed):
A few minor things, but this is looking good to me.

kerimoyle committed:
changes as per review

kerimoyle committed:
Merge branch 'develop' into i652_parse_print_duplicate_ids

kerimoyle committed:
Merge pull request #45 from cellml/develop

update from upstream

kerimoyle committed:
Merge branch 'develop' into i652_parse_print_duplicate_ids

kerimoyle committed:
Fix merge conflicts

kerimoyle commented:
Conflicts fixed.  Ready for review please @agarny @nickerso @hsorby

agarny commented on src/api/libcellml/component.h:
```diff
@@ -414,6 +416,16 @@ class LIBCELLML_EXPORT Component: public ComponentEntity, public ImportedEntity
+    /**
+     * @brief Set the import source of this component.
+     *
+     * If this component is already located in a Model instance, then the
```
Should be "... @c Model instance..." Same on the line below.

kerimoyle replied:
This is inconsistent everywhere ... and should be `@c ModelPtr` or `@c Model` ??

agarny replied:
My personal preference is for `@c Model`.

kerimoyle replied:
Will be addressed in #618

agarny commented on src/api/libcellml/importsource.h:
```diff
@@ -109,6 +117,161 @@ class LIBCELLML_EXPORT ImportSource: public Entity
+     * @brief Add a component to this import source item.
+     *
+     * The added @p component will be treated as an import, and
+     * will be sourced from the combination of url of this import source
```
"URL", not "url".

agarny replied:
Same elsewhere.

agarny commented on src/api/libcellml/importsource.h:
```diff
@@ -109,6 +117,161 @@ class LIBCELLML_EXPORT ImportSource: public Entity
+     * will be sourced from the combination of url of this import source
+     * item, and the component's own import reference.
+     *
+     * @param component The @c ComponentPtr to add.
```
"Component", not "ComponentPtr".

kerimoyle replied:
... but the parameter isn't a `Component`, it's a `ComponentPtr`?

agarny replied:
Same elsewhere...

agarny replied:
Actually, the documentation is not consistent. See the one for `removeComponent()`.

kerimoyle replied:
Will be addressed in #618

agarny commented on src/api/libcellml/importsource.h:
```diff
@@ -109,6 +117,161 @@ class LIBCELLML_EXPORT ImportSource: public Entity
+     * @brief Remove the component at the given @p index.
+     *
+     * Remove the component from this import source at the given @p index.
+     * @p index must be in the range [0, \#component).
```
"components", not "component" (I believe this is what we do in other places, i.e. use plural).

kerimoyle replied:
Done (somewhere...)

agarny commented on src/api/libcellml/importsource.h:
```diff
@@ -109,6 +117,161 @@ class LIBCELLML_EXPORT ImportSource: public Entity
+     * @p index must be in the range [0, \#component).
+     *
+     * @param index The index of the component to remove.
+     * @param setEmpty Whether or not to set the removed Component's
```
"@c Component"

agarny replied:
Same elsewhere.

kerimoyle replied:
Will be addressed in #618

agarny commented on src/api/libcellml/importsource.h:
```diff
@@ -109,6 +117,161 @@ class LIBCELLML_EXPORT ImportSource: public Entity
+     *
+     * @param index The index of the component to remove.
+     * @param setEmpty Whether or not to set the removed Component's
+     *        importSource to nullptr.  True by default.
```
We should something have "`nullptr`" rather than "nullptr".

agarny replied:
Same elsewhere.

agarny replied:
I.e. "@c nullptr" rather than "nullptr".

kerimoyle replied:
Will be addressed in #618

agarny commented on src/api/libcellml/importsource.h:
```diff
@@ -109,6 +117,161 @@ class LIBCELLML_EXPORT ImportSource: public Entity
+    /**
+     * @brief Remove all components from the import source.
+     *
+     * Note that components which are removed from the import will have their
```
Need a "proper" description of the method before that note.

kerimoyle replied:
I don't understand? Doesn't the brief comment say it?

agarny replied:
Same for `removeAllUnits()`.

agarny replied:
Sure, it does, but my understanding is that this section should contain a more detailed description, which is missing here.

kerimoyle replied:
e92e73bb

agarny commented on src/api/libcellml/importsource.h:
```diff
@@ -109,6 +117,161 @@ class LIBCELLML_EXPORT ImportSource: public Entity
+     * Note that components which are removed from the import will have their
+     * import source cleared.  They thus become locally
+     * concrete instances in the model, and will return @c false to the
+     * function isImport().
```
Need to properly reference `isImport()`.

kerimoyle replied:
I don't understand?

agarny replied:
Sorry, I meant something like `@sa isImport`?

agarny commented on src/api/libcellml/importsource.h:
```diff
@@ -109,6 +117,161 @@ class LIBCELLML_EXPORT ImportSource: public Entity
+     * @p index must be in the range [0, \#units).
+     *
+     * @param index The index of the units to remove.
+     * @param setEmpty Whether or not to set the removed Units'
```
"@c Units"

kerimoyle replied:
Will be addressed in #618

agarny commented on src/api/libcellml/model.h:
```diff
@@ -366,6 +366,89 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+    /**
+     * @brief Add an import source item to this model.
+     *
+     * Add import source by reference to the model. If the import
```
Would replace with "Add the @p importSource to the model" and maybe rephrase the rest of that paragprah accordingly.

kerimoyle replied:
TBD in #618

agarny commented on src/api/libcellml/model.h:
```diff
@@ -366,6 +366,89 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+     *
+     * The function will return @c false and no action is taken if:
+     *  - The @p imp pointer already exists in this model; or
+     *  - The @p imp is @c nullptr.
```
Replace "imp" with "importSource".

agarny replied:
Same below.

kerimoyle replied:
It has a @p for parameter, so needs to be the same as is listed in the declaration?

agarny replied:
... and elsewhere.

agarny replied:
Yes, I would expect the declaration to use `importSource` rather than `imp`, which in turn means that we should here use `@p importSource` rather than `@p imp`.

kerimoyle replied:
It's not great to use variable names which shadow function names, which is why I chose to use `imp` and not `importSource`?

agarny replied:
?? The method is called `addImportSource()`, so where is the problem?

kerimoyle replied:
If you declare a variable called `importSource` in the header, then you have a variable called `importSource` in the function.  Since we have a function called `importSource` in this class, then also having a variable with the same name isn't great?

agarny replied:
I have no problems with that, not least because this is exactly what is currently being done in `api/libcellml/importedentity.h` and in `api/libcellml/issue.h` where a method `importSource()` is defined and `importSource` is used as a parameter name for `setImportSource()`.

kerimoyle replied:
Heh, I would rather change those ones too then ... I would have thought that there would be a clang flag for this kind of thing? Why do we only use `-Wshadow` in gcc? I'll change it to a longer variable name, but I really don't like shadowing others, so it won't be `importSource`.

kerimoyle replied:
f85f512e

agarny commented on src/api/libcellml/model.h:
```diff
@@ -366,6 +366,89 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+    bool addImportSource(const ImportSourcePtr &imp);
+
+    /**
+     * @brief Get the number of import source items in the model.
```
Why "import source items" and not simply "import sources"?

agarny replied:
Same elsewhere.

kerimoyle replied:
Because - subtly - they're not the same thing.  We can have one import source, and many import source items.

agarny replied:
Oops, my bad. I stand corrected!

agarny commented on src/api/libcellml/model.h:
```diff
@@ -366,6 +366,89 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+     *
+     * Returns a reference to an import source at the given @p index.  If the @p index
+     * is not valid a @c nullptr is returned, the range of valid values for the
+     * index is [0, \#import sources).
```
I would replace wih "#importSources".

agarny replied:
Same elsewhere.

kerimoyle replied:
I don't understand the reason.  Elsewhere we use lowercase descriptions? It would be better to ditch the `#` altogether as it's weird to use this kind of shorthand here.  Better to spell it out properly, or provide a reference to the `importSourceCount()` function.

agarny replied:
Well, we do have "[0, \#equivalentVariables)" in `variable.h`. I am "fine" with whatever we go for, as long as we are being consistent and here we are not.

kerimoyle replied:
7a544255

agarny commented on src/api/libcellml/units.h:
```diff
@@ -491,6 +492,16 @@ class LIBCELLML_EXPORT Units: public NamedEntity, public ImportedEntity
+    /**
+     * @brief Set the import source of these units.
+     *
+     * If these units are already located in a Model instance, then the
```
"@c Model". Same below.

kerimoyle replied:
TBD in #618

agarny commented on src/bindings/interface/importsource.i:
```diff
@@ -31,6 +31,36 @@ unset).";
 "Create a copy of this import source.";
 
+%feature("docstring") libcellml::ImportSource::addComponent
+"Add a dependent Component to this ImportSource.";
```
I would remove "dependent".

agarny replied:
Also, would use "component" not "Component". (This is what we do elsewhere in our interfaces.) Same below.

kerimoyle replied:
Why? It has knock-on effects for the component, so it is a dependent component?

agarny replied:
Same with "ImportSource" which should be replaced with "import source".

agarny replied:
Just because the method is called `addComponent`. If it was called `addDependentComponent` (which I wouldn't want) then it would make sense for me to have "dependent", but not as it stands. Just my 2c.

kerimoyle replied:
Eh.  I'd hope that docstrings were allowed to have more info than just rewriting their function name!  I've changed it for now 7a544255 but hopefully #618 can improve what we provide to users.

agarny commented on src/bindings/interface/importsource.i:
```diff
@@ -31,6 +31,36 @@ unset).";
+"Retrieve the imported Component at the given index.";
+
+%feature("docstring") libcellml::ImportSource::addUnits
+"Add a dependent Units item to this ImportSource.";
```
"units", not "Units". Same below.

kerimoyle replied:
TBD in #618

agarny commented on src/bindings/interface/model.i:
```diff
@@ -76,6 +76,25 @@ determine the full path to the source model relative to this one.";
 "Fix variable interfaces throughout the model.";
 
+%feature("docstring") libcellml::Model::addImportSource
+"Manually add an ImportSource to the Model.";
```
"import source" and "model", not "ImportSource" and "Model". Same below.

kerimoyle replied:
TBD in #618

agarny commented on src/bindings/interface/model.i:
```diff
@@ -76,6 +76,25 @@ determine the full path to the source model relative to this one.";
+"Return the number of ImportSource items in the Model.";
+
+%feature("docstring") libcellml::Model::importSource
+"Retrieve an ImportSource by index or url.";
```
Reminder: "URL". Same below.

kerimoyle replied:
81328612

agarny commented on src/importsource.cpp:
```diff
@@ -14,12 +14,20 @@ See the License for the specific language governing permissions and
 */
 
-#include "libcellml/importsource.h"
+#include <algorithm> // Adding because Ubuntu build complained?
```
Remove comment.

kerimoyle replied:
81328612

agarny commented on src/importsource.cpp:
```diff
@@ -14,12 +14,20 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-#include "libcellml/importsource.h"
```
This shouldn't have been removed (this is how we do things now: add the `.cpp`'s corresponding `.h` file first, then system headers then API headers, with an empty line between each category), and an empty line added.

kerimoyle replied:
81328612

agarny commented on src/importsource.cpp:
```diff
@@ -14,12 +14,20 @@ See the License for the specific language governing permissions and
 
+#include "libcellml/component.h"
+#include "libcellml/importedentity.h"
+#include "libcellml/importsource.h"
```
Remove this include (since it should be included at the top).

kerimoyle replied:
81328612

agarny commented on src/importsource.cpp:
```diff
@@ -83,4 +96,211 @@ ImportSourcePtr ImportSource::clone() const
+    mPimpl->removeItem(it);
+
+    if (setEmpty) {
+        ImportSourcePtr empty = nullptr;
+        component->setImportSource(empty);
```
As previously agreed/discussed, this should be replaced with `component->setImportSource(nullptr);`. Same below.

kerimoyle replied:
It's not possible to do this (as I think we did discuss last time!) The compiler complains about the non-const lvalue reference.

agarny replied:
Oh, that's because there is no `const` in `setImportSource()`. Ok, got it.

kerimoyle committed:
some changes from review

agarny commented on src/importsource.cpp:
```diff
@@ -83,4 +96,211 @@ ImportSourcePtr ImportSource::clone() const
+    }
+    // Remove the item from the imports.
+    mImports.erase(it);
+}
```
Definitely not keen on all that casting, not least since `size()` returns a `size_t`, so to cast it to an `int` is "risky" although we are very unlikely to ever have more than 2,147,483,647 components/units. Anyway, I would rewrite this method to read:
```
void ImportSource::ImportSourceImpl::removeItem(std::vector<ImportedEntityWeakPtr>::iterator &it)
{
    // Find current index for the item to remove.
    auto index = size_t(std::distance(mImports.begin(), it));
    auto itU = std::find(mUnits.begin(), mUnits.end(), index);
    if (itU != mUnits.end()) {
        mUnits.erase(itU);
    } else {
        auto itC = std::find(mComponents.begin(), mComponents.end(), index);
        if (itC != mComponents.end()) {
            mComponents.erase(itC);
        }
    }

    // Iterate through components index vector and decrement all indices above the item's index.
    static const size_t MAX_SIZE_T = std::numeric_limits<size_t>::max();
    for (size_t i = mComponents.size() - 1; (i != MAX_SIZE_T) && (mComponents[i] > index); --i) {
        --mComponents[i];
    }
    // Loop through units vector and decrease its indices too.
    for (size_t i = mUnits.size() - 1; (i != MAX_SIZE_T) && (mUnits[i] > index); --i) {
        --mUnits[i];
    }
    // Remove the item from the imports.
    mImports.erase(it);
}
```

kerimoyle replied:
a8276f79

kerimoyle commented:
Thanks for the review @agarny.  I've made some of the changes as suggested, but I'm not prepared to spend time doing the docstrings until we've got some kind of agreement with @hsorby and @nickerso about what they should be everywhere.  And even then, there is so much inconsistency throughout I think they should be addressed in a separate PR which fixes up everything properly.

agarny commented on src/parser.cpp:
```diff
@@ -1175,6 +1176,8 @@ void Parser::ParserImpl::loadImport(const ImportSourcePtr &importSource, const M
         if (childNode->isCellmlElement("component")) {
             ComponentPtr importedComponent = Component::create();
+            // importedComponent->setImportSource(importSource);
+
```
Delete.

kerimoyle replied:
dc8c5607

agarny commented on src/parser.cpp:
```diff
@@ -1198,6 +1201,7 @@ void Parser::ParserImpl::loadImport(const ImportSourcePtr &importSource, const M
             }
         } else if (childNode->isCellmlElement("units")) {
             UnitsPtr importedUnits = Units::create();
+            // importedUnits->setImportSource(importSource);
```
Delete.

kerimoyle replied:
dc8c5607

kerimoyle committed:
rewrite removeItem

agarny commented on tests/model/component_import.cpp:
```diff
@@ -329,3 +319,62 @@ TEST(ComponentImport, complexImportAndParse)
+    // Add a new import source with a url that's already in the list:
+    // NB This has changed so that the new import source
+    // object will trigger an additional import block, even though the
+    // url (and therefore the imported object) already exists.
```
URL

kerimoyle replied:
done already

kerimoyle committed:
removed commented functions

agarny commented on tests/model/units_import.cpp:
```diff
@@ -215,3 +215,65 @@ TEST(UnitsImport, importModifyAndParse)
+
+    EXPECT_EQ(size_t(1), model->importSourceCount());
+    EXPECT_EQ(imp1, model->importSource(0));
+    // EXPECT_EQ(imp1, model->importSource(url1));
```
Delete.

agarny replied:
Same below.

kerimoyle replied:
2f009081

agarny commented on tests/import_source/import_source.cpp:
```diff
@@ -0,0 +1,401 @@
+        auto c2 = m2->component(index);
+        compareComponent(c1, c2, m2);
+    }
+}
```
Shouldn't these be in `test_utils`? Might come handy for some other tests at some point?

kerimoyle replied:
2f009081

agarny commented on tests/import_source/import_source.cpp:
```diff
@@ -0,0 +1,401 @@
+    EXPECT_EQ(size_t(2), model->importSourceCount());
+    EXPECT_EQ(imp2, model->importSource(1));
+
+    // Add another with a new url:
```
URL

kerimoyle replied:
2f009081

agarny commented on tests/import_source/import_source.cpp:
```diff
@@ -0,0 +1,401 @@
+
+    // Add the other units so it's an import.
+    units2->setImportSource(imp);
+
```
Delete empty line.

kerimoyle replied:
2f009081

agarny commented on tests/import_source/import_source.cpp:
```diff
@@ -0,0 +1,401 @@
+    EXPECT_TRUE(component->isImport());
+    EXPECT_TRUE(units->isImport());
+    EXPECT_EQ(size_t(1), imp1->componentCount());
+    EXPECT_EQ(size_t(1), imp1->componentCount());
```
Did you mean `EXPECT_EQ(size_t(1), imp1->unitsCount());`?

kerimoyle replied:
2f009081

agarny commented on tests/import_source/import_source.cpp:
```diff
@@ -0,0 +1,401 @@
+    component->setImportSource(imp2);
+    units->setImportSource(imp2);
+    EXPECT_EQ(size_t(0), imp1->componentCount());
+    EXPECT_EQ(size_t(0), imp1->componentCount());
```
`EXPECT_EQ(size_t(0), imp1->unitsCount());`

kerimoyle replied:
2f009081

agarny commented on tests/import_source/import_source.cpp:
```diff
@@ -0,0 +1,401 @@
+    EXPECT_EQ(size_t(0), imp1->componentCount());
+    EXPECT_EQ(size_t(0), imp1->componentCount());
+    EXPECT_EQ(size_t(1), imp2->componentCount());
+    EXPECT_EQ(size_t(1), imp2->componentCount());
```
`EXPECT_EQ(size_t(1), imp2->unitsCount());`

kerimoyle replied:
2f009081

agarny reviewed (changes requested):
Various minor changes, but otherwise looks good.

agarny commented:
> Thanks for the review @agarny. I've made some of the changes as suggested, but I'm not prepared to spend time doing the docstrings until we've got some kind of agreement with @hsorby and @nickerso about what they should be everywhere. And even then, there is so much inconsistency throughout I think they should be addressed in a separate PR which fixes up everything properly.

Sure, happy for this to be done in a separate PR.

kerimoyle committed:
changes as per review.

kerimoyle committed:
small changes to docstrings

kerimoyle committed:
Update importsource.i

agarny commented on src/api/libcellml/model.h:
```diff
@@ -420,9 +420,7 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
-     * @brief Remove the import source with the given pointer.
-     *
-     * Remove the import source with the pointer @p imp.
+     * @brief Remove the import source at the given reference.
```
Why has the "detailed" description been removed?

kerimoyle replied:
Because it was the same as the brief one.  No point in duplicating it.

agarny replied:
Well, that breaks consistency (and not sure how it will look like in the documentation), but I have voiced my view, so I will leave it to @hsorby and @nickerso to comment.

nickerso replied:
No point having a full description that is the same as the brief description (doxygen doesn't care, it will use the brief as needed). In my view the full description is only needed if the brief description is not sufficient.

agarny replied:
We are in a democracy, so... fine. :)

kerimoyle committed:
renamed imp variable to importSrc

nickerso commented:
> Sure, happy for this to be done in a separate PR.

agree.

agarny reviewed (approved):
I really don't like commit f85f512. It's not consistent with the rest of our codebase and that's also the reason I originally advocated parameter names to start with `p` to avoid problems like this, and also to make it clear whether a variable is a member variable (e.g. `mVariable`), a parameter variable (`pVariable`) or a local variable (`variable`). Anyway...
